python/lsst/ctrl/bps/generic_workflow.py

```diff
@@ -581,3 +638,102 @@ def validate(self):
         # Make sure a directed acyclic graph
         assert nx.algorithms.dag.is_directed_acyclic_graph(self)
+
+    def add_workflow_source(self, workflow):
+        """Add given workflow as new source to this workflow.
+
+        Parameters
+        ----------
+        workflow : `lsst.ctrl.bps.GenericWorkflow`
+        """
+        # Find source nodes in self.
+        self_sources = [n for n in self if self.in_degree(n) == 0]
+        _LOG.debug("self_sources = %s", self_sources)
+
+        # Find sink nodes of workflow.
+        new_sinks = [n for n in workflow if workflow.out_degree(n) == 0]
+        _LOG.debug("new sinks = %s", new_sinks)
+
+        # Add new workflow nodes to self graph and make new edges.
+        self.add_nodes_from(workflow.nodes(data=True))
+        self.add_edges_from(workflow.edges())
+        for source in self_sources:
+            for sink in new_sinks:
+                self.add_edge(sink, source)
+
+        # Files are stored separately so copy them.
+        for job_name in workflow:
+            self.add_job_inputs(job_name, workflow.get_job_inputs(job_name, data=True))
+            self.add_job_outputs(job_name, workflow.get_job_outputs(job_name, data=True))
+            self.add_executable(workflow.get_job(job_name).executable)
+
+    def add_final(self, final):
+        """Add special final job/workflow to the generic workflow.
+
+        Parameters
+        ----------
+        final : `lsst.ctrl.bps.GenericWorkflowJob` or
+                `lsst.ctrl.bps.GenericWorkflow`
+            Information needed to execute the special final job(s), the
+            job(s) to be executed after all jobs that can be executed
+            have been executed regardless of exit status of any of the
+            jobs.
+        """
+        if not isinstance(final, GenericWorkflowJob) and not isinstance(final, GenericWorkflow):
+            raise TypeError("Invalid type for GenericWorkflow final ({type(final)})")
+
+        self._final = final
+        if isinstance(final, GenericWorkflowJob):
+            self.add_executable(final.executable)
+
+    def get_final(self):
+        """Return job/workflow to be executed after all jobs that can be
+        executed have been executed regardless of exit status of any of
+        the jobs.
+
+        Returns
+        -------
+        final : `lsst.ctrl.bps.GenericWorkflowJob` or
+                `lsst.ctrl.bps.GenericWorkflow`
+            Information needed to execute final job(s).
+        """
+        return self._final
+
+    def add_executable(self, executable):
+        """Add executable to workflow's list of executables.
+
+        Parameters
+        ----------
+        executable : `lsst.ctrl.bps.GenericWorkflowExec`
+            Executable object to be added to workflow.
+        """
+        if executable is not None:
+            self._executables[executable.name] = executable
+        else:
+            _LOG.warning("add_executable the executable is None")
```

Comment:
Can we change the message to something descriptive, for example "executable not specified (None); cannot add to the workflow's list of the executables"?